### PR TITLE
WIP/MESOS: Fix endpoint related e2e tests on Mesos

### PR DESCRIPTION
**Pull request by sttts**
Before this PR endpoints were validated by container IP and port in a number of e2e tests.
Depending on the endpoint controller logic neither of the two must match for a valid endpoint (e.g. in a Mesos setup).

This PR
- fixes the service.go e2e tests to compare the pod ID of endpoint instead of the IP
- makes sure the compared port is the actual container port, not the host port.

The **short-term solution** to recover the container ports from an endpoint in the case of mesos/docker clusters are annotations on the endpoints. These are applied by the Mesos endpoint conroller to be able to lookup the container port which usually is not part of the EndpointPort struct.

The **proposed long-term solution** is to add a ContainerPort field to the EndpointPort struct. This should default to the Port value, but be set to the actually container port by endpoint controller like that of Mesos.

This PR is on-top of https://github.com/GoogleCloudPlatform/kubernetes/pull/10390.

### CODE REVIEW

File: test/e2e/service.go [hunk 1]
```diff
@@ -1021,60 +1023,82 @@ func validateUniqueOrFail(s []string) {
...
 		for _, port := range ss.Ports {
 			for _, addr := range ss.Addresses {
-				Logf("Found IP %v and port %v", addr.IP, port.Port)
-				if _, ok := m[addr.IP]; !ok {
-					m[addr.IP] = make([]int, 0)
+				containerPort := port.Port
+				hostPort := port.Port
+
+				// use endpoint annotations to recover the container port in a Mesos setup
+				// TODO(sttts): extend EndpointPort by ContainerIP, defaulting to Port
```
Comment (jdef) [inline test/e2e/service.go]: you really mean "add ContainerPort to EndpointPort, defaults to Port", right?

File: test/e2e/service.go [hunk 2]
```diff
@@ -1021,60 +1023,82 @@ func validateUniqueOrFail(s []string) {
...
+	for _, ss := range endpoints.Subsets {
 		for _, port := range ss.Ports {
 			for _, addr := range ss.Addresses {
-				Logf("Found IP %v and port %v", addr.IP, port.Port)
-				if _, ok := m[addr.IP]; !ok {
-					m[addr.IP] = make([]int, 0)
+				containerPort := port.Port
+				hostPort := port.Port
+
+				// use endpoint annotations to recover the container port in a Mesos setup
```
Comment (jdef) [inline test/e2e/service.go:1035]: can you add a reference to the endpoint controller package that adds the annotations?

**Comment by jdef** [comment]
I'm not sure that the _long term plan_ is viable: the core team strongly dislikes our non-IP-per-pod approach. Trying to push a ContainerPort into EndpointPort as part of a workaround for something that's already viewed as a kludgy workaround isn't going to fly.

That said I think the annotations you've added are fine: they're isolated to our already-hacky endpoint controller and there's minimal impact on the test suite.

File: test/e2e/service.go (hunk 1 above)
Comment (sttts) [inline test/e2e/service.go]: yes. A bit like the Port+targetPort in the services, but for container port and host port.

File: test/e2e/service.go (hunk 2 above)
Comment (sttts) [inline test/e2e/service.go:1035]: done.

**Comment by sttts** [comment]
I have the same expectation about the EndpointPort plan. Let's see. 

As you say: it's a pretty local change now, not like before modifying a number of e2e tests to named ports.

**Pull request closed without merging**